### PR TITLE
WT-4026 Add implementation for existing file extension configuration API

### PR DESCRIPTION
**Pull request by nehakhatri5**
File extend API on Windows(setEndOfFile) could be slow at times based on MSDN API
documentation hence disable it.

### CODE REVIEW

**Comment by agorrod** [comment]
Thanks @nehakhatri5 - let's do some testing with this. I think the final version will need to have a bit more nuance - we also use the fh_extend interface to truncate data files after a compact - and I'm reluctant to give that up. We might also want to make this choice configurable.

**Comment by sueloverso** [comment]
I have read the ticket and certainly this change is very simple. However is there any information about impact in the areas @agorrod mentioned such as compact? We already have the `file_extend` configuration knob on `wiredtiger_open`. Perhaps documentation could be used for this instead of the code change. Does `file_extend=(data=0)` work or show the same performance as removing it entirely?

File: src/include/wiredtiger.in [hunk 1]
```diff
@@ -2837,10 +2837,12 @@ struct __wt_connection {
...
  * default empty.}
- * @config{file_extend, file extension configuration.  If set\, extend files of
- * the set type in allocations of the set size\, instead of a block at a time as
+ * @config{file_extend, file extension configuration.  If set for data files\,
+ * extend those in allocations of the set size\, instead of a block at a time as
  * each new block is written.  For example\,
- * <code>file_extend=(data=16MB)</code>., a list\, with values chosen from the
+ * <code>file_extend=(data=16MB)</code>. Disables log file extension when set to
+ * -1 for log file.  The default is to extend the log file in allocations of the
+ * <code>log_file_max</code> size., a list\, with values chosen from the
```
Comment (sueloverso) [inline src/include/wiredtiger.in]: I have a couple comments on making this configurable via `file_extend`:
1. It feels weird to make it specific only to log extension. I think someone may want to turn it off for either type of file.
2. Wouldn't setting it to 0 turn it off? How is 0 different than -1 here? What happens today if the user sets it to 0? I think they're the same, and if so, using 0 feels more natural.

File: src/log/log.c [hunk 2]
```diff
@@ -673,6 +673,10 @@ __log_prealloc(WT_SESSION_IMPL *session, WT_FH *fh)
 		return (__log_zero(session, fh,
 		    WT_LOG_END_HEADER, conn->log_file_max));
 
+	/* If configured to not extend the file, we're done. */
+	if (conn->log_extend_len == -1)
```
Comment (sueloverso) [inline src/log/log.c]: This may change of you decide to use 0. Also consider adding for data files too - or does that fall out if someone uses 0.

File: src/include/wiredtiger.in (hunk 1 above)
Comment (nehakhatri5) [reply src/include/wiredtiger.in]: Ans 1. Currently, log files get extended irrespective of `file_extend` config.  See here:
https://github.com/wiredtiger/wiredtiger/blob/333460997573558235254cb7e84dad765bb844a7/src/log/log.c#L680]
So `file_extend` config is working differently for data files and log files, hence we are making the change specific only to log files.

Ans 2 : Setting it to 0 doesn't turn it off for log files. The default is 0, which means data file extension is off, but log file extension is on. So they are not the same. Hence to disable log file extension, set it to -1.

File: src/log/log.c (hunk 2 above)
Comment (nehakhatri5) [reply src/log/log.c]: If you agree with my comment above, then we would use -1 to disable log file extention.

**Comment by sueloverso** [comment]
Since this is a user-facing API change, I find the usage confusing and think we should have a more general discussion.
* Does log=(prealloc) factor into this? I.e. does `prealloc` override, interfere with or otherwise interact with `file_extend`? It seems like pre-allocating could just create the files and write/sync the header without extending so maybe not.
* It does not feel natural that `data=0` means don't extend data files and `log=-1` would mean don't extend. That makes them inconsistent with each other. Perhaps the default for `file_extend=(log=...)` needs to be something other than 0 so that we usually extend log files and then use `log=0` to turn it off?

**Comment by lukech** [comment]
Do we need to update "docs/upgrading.dox" to reflect this API change?

**Comment by agorrod** [comment]
I've added follow up on the WT-4026 JIRA ticket.

**Comment by agorrod** [comment]
Thanks @nehakhatri5 this change lgtm now - @sueloverso please take another look

**Comment by nehakhatri5** [comment]
@sueloverso Please take look. If the API looks OK, I will update the upgrading.dox.

**Comment by nehakhatri5** [comment]
Thanks @sueloverso, I have updated the code based on your suggestions. I will also briefly discuss with @Alex on Monday, about the precise range of allowed values for log file extension size. After that I will put changes for review.

**Comment by nehakhatri5** [comment]
@sueloverso I agree with all of the suggestions and have updated the code for those. Please take a look and let me know.

File: src/conn/conn_api.c [hunk 3]
```diff
@@ -2626,14 +2626,35 @@ wiredtiger_open(const char *home, WT_EVENT_HANDLER *event_handler,
...
 	WT_ERR(__wt_config_gets(session, cfg, "file_extend", &cval));
 	for (ft = file_types; ft->name != NULL; ft++) {
+		/*
+		 * An unset log extend length is an indicator to set it
+		 * with configured maximum log file size in case of the default
+		 * configuration. Update the log extend length with actual
+		 * desired size when maximum log file size gets initialized as
+		 * part log server initialization.
+		 */
+		conn->log_extend_len = WT_CONFIG_UNSET;
```
Comment (sueloverso) [inline src/conn/conn_api.c]: I think this needs to be above the for-loop. Otherwise `file_extend=(log=1M,data=16M)` would overwrite the set `log_extend_len` when it processed the `data=16M` part on the 2nd time through the loop.

File: src/conn/conn_api.c [hunk 4]
```diff
@@ -2626,14 +2626,35 @@ wiredtiger_open(const char *home, WT_EVENT_HANDLER *event_handler,
 
 	WT_ERR(__wt_config_gets(session, cfg, "file_extend", &cval));
 	for (ft = file_types; ft->name != NULL; ft++) {
+		/*
+		 * An unset log extend length is an indicator to set it
+		 * with configured maximum log file size in case of the default
+		 * configuration. Update the log extend length with actual
+		 * desired size when maximum log file size gets initialized as
+		 * part log server initialization.
+		 */
```
Comment (sueloverso) [inline src/conn/conn_api.c]: I suggest rewording this comment:
* If the log extend length is not set use the default of the configured maximum
* log file size. That size is not known until it is initialized as part of the log
* server initialization.

File: src/conn/conn_api.c [hunk 5]
```diff
@@ -2626,14 +2626,35 @@ wiredtiger_open(const char *home, WT_EVENT_HANDLER *event_handler,
...
 		if (ret == 0) {
 			switch (ft->flag) {
 			case WT_DIRECT_IO_DATA:
 				conn->data_extend_len = sval.val;
 				break;
 			case WT_DIRECT_IO_LOG:
-				conn->log_extend_len = sval.val;
+				/*
+				 * In the case of the "file_extend=(log=)",
+				 * unset the log extend length.
```
Comment (sueloverso) [inline src/conn/conn_api.c]: I suggest a minor change to this comment:
* When using "file_extend=(log=)", the val returned is 1.
* Unset the log extend length in that case to use the default.

File: src/conn/conn_log.c [hunk 6]
```diff
@@ -277,6 +277,15 @@ __logmgr_config(
 	if (!reconfig) {
 		WT_RET(__wt_config_gets(session, cfg, "log.file_max", &cval));
 		conn->log_file_max = (wt_off_t)cval.val;
+		/*
+		 * With the default log file extend configuration or log file
+		 * extension size larger than configured maximum log file size,
+		 * set the log file extension size to configured maximum log
```
Comment (sueloverso) [inline src/conn/conn_log.c]: minor nit: "...to the configured maximum log..."

File: src/include/log.h [hunk 7]
```diff
@@ -87,6 +87,12 @@ union __wt_lsn {
 #define	WT_LOGC_KEY_FORMAT	WT_UNCHECKED_STRING(III)
 #define	WT_LOGC_VALUE_FORMAT	WT_UNCHECKED_STRING(qIIIuu)
 
+/*
+ * Size range for the log files.
+ */
+#define	WT_LOG_FILE_MAX ((uint64_t)2 * WT_GIGABYTE)
+#define	WT_LOG_FILE_MIN (100 * 1024)
```
Comment (sueloverso) [inline src/include/log.h]: Minor nit: change `1024` to `WT_KILOBYTE`.

File: test/suite/test_config07.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+        logs = fnmatch.filter(os.listdir('.'), "*Prep*")
+        f = logs[-1]
+        file_size = os.stat(f).st_size
+        self.assertEqual(size, file_size)
+
+    def test_log_extend(self):
+        self.conn.close()
+        msg = '/invalid log extend length/'
+
+        config = 'log=(enabled,file_max=100K),file_extend=' + self.log_extend_len
```
Comment (sueloverso) [inline test/suite/test_config07.py]: I think you should use `file_max=1M` so that you can test smaller and larger (but within valid range).

File: test/suite/test_config07.py [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+import wiredtiger, wttest
+from wtscenario import make_scenarios
+
+# test_config07.py
+#    Test that log files extend as configured and as documented.
+class test_config07(wttest.WiredTigerTestCase):
+    uri = "table:test"
+    entries = 100000
+    K = 1024
+
```
Comment (sueloverso) [inline test/suite/test_config07.py]: I suggest replacing this variable with  `log_size = ` and then you can use the same log size in the `expected_log_size` definitions and the connection configuration so that if we want to test other log sizes, only one place needs to change.

File: test/suite/test_config07.py [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+    extend_len = [
+        ('default', dict(log_extend_len='()', expected_log_size = 100 * K)),
+        ('empty', dict(log_extend_len='(log=)', expected_log_size = 100 * K)),
+        ('disable', dict(log_extend_len='(log=0)', expected_log_size = 128)),
+        ('100K', dict(log_extend_len='(log=100K)', expected_log_size = 100 * K)),
+        ('too_small', dict(log_extend_len='(log=20K)', expected_log_size = None)),
+        ('too_large', dict(log_extend_len='(log=20G)', expected_log_size = None)),
+        ('larger_than_log_file_size', dict(log_extend_len='(log=20M)',
+                                          expected_log_size = 100 * K)),
```
Comment (sueloverso) [inline test/suite/test_config07.py]: I added a comment below to change the log file size so that we can test valid ranges larger and smaller than the `file_max`. So I think you should add a `1M` scenario above.

My earlier comment about the bug in the for-loop made me think that it would be useful to have a scenario of:
`dict(log_extend_len='(log=100K,data=16M)',` to confirm my suspicion that it gets overwritten and verify the fix.

File: src/conn/conn_api.c (hunk 3 above)
Comment (nehakhatri5) [reply src/conn/conn_api.c]: It's a good point. Though the "file_types" array defined in this file ensures that data file config always gets parsed before the log file config. That way `log_extend_len` would not get overwritten with` file_extend=(log=1M,data=16M)`. However it makes sense to initialise the log_extend_len above the loop. I have made the change.

File: test/suite/test_config07.py (hunk 9 above)
Comment (nehakhatri5) [reply test/suite/test_config07.py]: I may not be able to replace `K=1024` with `log_size =` but I have added a new variable `log_size`. This is because `K` will be required by the `expected_log_size` definitions of the valid cases like `log_extend_len='(log=100K)'`.

File: test/suite/test_config07.py (hunk 8 above)
Comment (nehakhatri5) [reply test/suite/test_config07.py]: Done.

File: test/suite/test_config07.py (hunk 10 above)
Comment (nehakhatri5) [reply test/suite/test_config07.py]: Done.

File: src/conn/conn_log.c (hunk 6 above)
Comment (nehakhatri5) [reply src/conn/conn_log.c]: Done.

File: src/conn/conn_api.c (hunk 5 above)
Comment (nehakhatri5) [reply src/conn/conn_api.c]: Done.

File: src/conn/conn_api.c (hunk 4 above)
Comment (nehakhatri5) [reply src/conn/conn_api.c]: Done.

File: src/include/log.h (hunk 7 above)
Comment (nehakhatri5) [reply src/include/log.h]: Done.
